src/model_rng.cpp

```diff
@@ -256,13 +256,22 @@ void bs_model::log_density(bool propto, bool jacobian, const double* theta_unc,
   if (propto) {
     // need to have vars, otherwise the result is 0 since everything is
     // treated as a constant
-    Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> params_unc_var(
-        params_unc);
-    if (jacobian) {
-      *val = model_->log_prob_propto_jacobian(params_unc_var, outstream).val();
-    } else {
-      *val = model_->log_prob_propto(params_unc_var, outstream).val();
+    try {
+      Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> params_unc_var(
+          params_unc);
+      if (jacobian) {
+        *val
+            = model_->log_prob_propto_jacobian(params_unc_var, outstream).val();
+      } else {
+        *val = model_->log_prob_propto(params_unc_var, outstream).val();
+      }
+    } catch (...) {
+      // because we created vars on the stack, we need to recover memory
+      stan::math::recover_memory();
+      throw;  // re-caught by top level exception logic
     }
+    // also recover memory if no exception was thrown
+    stan::math::recover_memory();
   } else {
     if (jacobian) {
```

Comment:
The try/catch needs to be one level up to handle the else case, too.  Anything that calls log prob and implicates autodiff.  I think these do.  But let me know if there's a reason the next calls won't use autodiff stack.

Reply:
params_unc is just a vector of doubles, so there won’t be any autodiff in the overload of log_prob it ends up calling.

I actually used the propto=False case to confirm that the memory leak was originally specific to this branch of things, so it should be good to go